src/arch/riscv/irqc/aia/vaplic.c

```diff
@@ -0,0 +1,1057 @@
+    spin_lock(&virqc->lock);
+    if (new_val != 0 && new_val < APLIC_MAX_INTERRUPTS && 
+        !get_bit_from_reg(virqc->setie[new_val/32], new_val)) {
+        set_bit_from_reg(&virqc->setie[new_val/32], new_val%32);
+        if(vaplic_get_hw(vcpu,new_val)){
+            aplic_set_ienum(new_val);
+        } else {
+            vaplic_update_hart_line(vcpu);
+        }
+    }
+    spin_unlock(&virqc->lock);
+}
+
+/**
+ * @brief Clear the enabled bits for interrupts [32*reg:(32*reg)+31]
+ * 
+ * @param vcpu virtual cpu 
+ * @param reg  regiter index
+ * @param new_val value with intp to be cleared per bit
+ */
+static void vaplic_set_clrie(struct vcpu *vcpu, uint8_t reg, uint32_t new_val){
+    struct virqc *virqc = &vcpu->vm->arch.virqc;
+    spin_lock(&virqc->lock);
+    if (reg < APLIC_NUM_CLRIx_REGS && 
+        vaplic_get_setie(vcpu, reg) != new_val) {
+        if (reg == 0) new_val &= 0xFFFFFFFE;
+        virqc->setie[reg] &= ~new_val;
+        for(int i = 0; i < APLIC_MAX_INTERRUPTS/APLIC_NUM_CLRIx_REGS; i++){
+            if(vaplic_get_hw(vcpu,i)){
+                if(!get_bit_from_reg(virqc->setie[reg], i) && ((new_val >> i) & 1)){
+                    aplic_set_clrienum(i);
+                }
+            } else {
+                vaplic_update_hart_line(vcpu);
+            }
+        }
+    }
+    spin_unlock(&virqc->lock);
+}
+
+/**
+ * @brief Clear the enabled bits for a given interrupt
+ * 
+ * @param vcpu virtual cpu
+ * @param new_val value w/ the interrupt source number
+ */
+static void vaplic_set_clrienum(struct vcpu *vcpu, uint32_t new_val){
+    struct virqc *virqc = &vcpu->vm->arch.virqc;
+    spin_lock(&virqc->lock);
+    if (new_val != 0 && new_val < APLIC_MAX_INTERRUPTS && 
+        get_bit_from_reg(virqc->setie[new_val/32], new_val)) {
+        clr_bit_from_reg(&virqc->setie[new_val/32], new_val%32);
+        if(vaplic_get_hw(vcpu,new_val)){
+            aplic_set_clrienum(new_val);
+        } else {
+            vaplic_update_hart_line(vcpu);
+        }
+    }
+    spin_unlock(&virqc->lock);
+}
+
+/**
+ * @brief Write to target register of a given interrupt
+ * 
+ * @param vcpu 
+ * @param intp_id interrupt id to write to
+ * @param new_val value to write to target
+ */
+static void vaplic_set_target(struct vcpu *vcpu, irqid_t intp_id, uint32_t new_val){
+    struct virqc *virqc = &vcpu->vm->arch.virqc;
+    uint32_t hart_index = (new_val >> APLIC_TARGET_HART_IDX_SHIFT) & APLIC_TARGET_HART_IDX_MASK;
+    cpuid_t pcpu_id = vm_translate_to_pcpuid(vcpu->vm, hart_index);
+
+    spin_lock(&virqc->lock);
+    if(pcpu_id == INVALID_CPUID){
+        /** If the hart index is invalid, make it vcpu = 0 
+         *  and read the new pcpu.
+         *  Software should not write anything other than legal 
+         *  values to such a field */
+        pcpu_id = vm_translate_to_pcpuid(vcpu->vm, 0);
+    }
+    /** Write the physical CPU in hart index */
+    new_val &= APLIC_TARGET_IPRIO_MASK;
```

Comment:
You need to check if this value is 0. If so, you need to set it to one (according to the spec).

Reply:
This is already fixed.